pkg/app/api/grpcapi/web_api.go

```diff
@@ -599,6 +602,47 @@ func (a *WebAPI) validatePipedBelongsToProject(ctx context.Context, pipedID, pro
 	return nil
 }
 
+func (a *WebAPI) ListUnregisteredApplications(ctx context.Context, _ *webservice.ListUnregisteredApplicationsRequest) (*webservice.ListUnregisteredApplicationsResponse, error) {
+	claims, err := rpcauth.ExtractClaims(ctx)
+	if err != nil {
+		a.logger.Error("failed to authenticate the current user", zap.Error(err))
+		return nil, err
+	}
+
+	// Collect all apps that belong to the project.
+	key := makeUnregisteredAppsCacheKey(claims.Role.ProjectId)
+	c := rediscache.NewHashCache(a.redis, key)
+	// pipedToApps assumes to be a map["piped-id"][]*model.ApplicationInfo
```

Comment:
The returned data will not be decoded to this `*model.ApplicationInfo` automatically to be ready to use since Redis doesn't have any way to know how the data was formatted. It just stored binary data.
So I think we have to encode (e.g. by JSON) the list before [storing](https://github.com/pipe-cd/pipe/blob/master/pkg/cache/rediscache/hashcache.go#L63) into Redis and decode them after [getting back](https://github.com/pipe-cd/pipe/blob/master/pkg/cache/rediscache/hashcache.go#L95) from Redis.

Reply:
Seriously? I am sorry, I had neglected to do my research. I assumed that `redigo` encodes/decodes using some kind of ways (such as gob). Let me look into how `redigo` encodes and re-apply new encoding way for us.